### PR TITLE
Fix VM restart during eviction when EvictionStrategy=LiveMigrate

### PR DESCRIPTION
**Pull request by davidvossel**
The issue here is that when a migration fails, the target pod isn't always immediately terminated. It can stick around for the duration of the virt-launcher timeout period which is 5 minutes.

If during that 5 minutes a new migration occurs, we run into a situation where a VMI might have > 2 pods active for the VMI at once. This invalidates our PDB, which can result in the VMI getting evicted even though we had EvictionStrategy=LiveMigrate

To make this work as reliably as possible, 3 changes are introduced to address this.

1. The eviction controller will never attempt to migrate a VMI if more than 1 pods are running already for the VMI
2. The migration controller will wait on creating the target pod until only 1 pod is running for a VMI.
3. A failed migration now signals the corresponding target pod to exit early if it's still waiting on the qemu process to appear.
4. The migration controller now finalizes the vmi's migration state if the target pod fails before handler starts the migration.

Items 2 and 3 prevent the PDB from being invalidated
Item 1 is an optimization that ensures the eviction controller will always attempt to act on VMIs which are immediately migratable. 
Item 4 ensures the VMI's migration state will move to a finalized state in the event that the target pod fails virt handler ever attempts to migrate the source.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1946441 

```release-note
Fixes VM restart during eviction when EvictionStrategy=LiveMigrate
```

### CODE REVIEW

**Comment by davidvossel** [comment]
/hold

i need to investigate some CI failures

**Comment by davidvossel** [comment]
/hold cancel

I figured out what was going on. It's addressed now. Tests should be passing reliably now.

**Comment by vladikr** [comment]
It all looks good to me. 
/approve

One question is about the `DownwardAPI`... I'm just curious why did you choose to implement it this way, instead of just letting virt-handler add an environment variable to virt-launcher?
It would save the additional mount.

**Comment by davidvossel** [comment]
> One question is about the DownwardAPI... I'm just curious why did you choose to implement it this way, instead of just letting virt-> handler add an environment variable to virt-launcher?
> It would save the additional mount.

yeah, i went back and forth on this. Ultimately, my thought process was that I was uncertain if I could reliably signal the launcher to exit early from the target virt-handler in the event that a migration fails.

There's just so much going on with the virt-handler side. For instance, from virt-handler, how would i know the pod i'm signalling to shutdown is the pod from the migration that previously failed, or perhaps a new pod for a new in-flight migration that is just booting up?

Virt-handler doesn't have that mapping of a migration object to a pod, but the migration controller does.

File: cmd/virt-launcher/virt-launcher.go [hunk 1]
```diff
@@ -218,27 +218,36 @@ func initializeDirs(virtShareDir string,
...
+
+	annotationsFile := filepath.Join(virtShareDir, "podinfo", "annotations")
+
+	for {
+		select {
+		case <-ticker:
+			panic(fmt.Errorf("timed out waiting for domain to be defined"))
+		case <-checkEarlyExit:
+			if util.HasEarlyExitAnnotation(annotationsFile) {
+				panic(fmt.Errorf("received early exit sinal via annotation at path %s before domain was detected", annotationsFile))
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply cmd/virt-launcher/virt-launcher.go]: fixed

**Comment by vladikr** [comment]
> > One question is about the DownwardAPI... I'm just curious why did you choose to implement it this way, instead of just letting virt-> handler add an environment variable to virt-launcher?
> > It would save the additional mount.
> 
> yeah, i went back and forth on this. Ultimately, my thought process was that I was uncertain if I could reliably signal the launcher to exit early from the target virt-handler in the event that a migration fails.
> 
> There's just so much going on with the virt-handler side. For instance, from virt-handler, how would i know the pod i'm signalling to shutdown is the pod from the migration that previously failed, or perhaps a new pod for a new in-flight migration that is just booting up?

These wouldn't have the same migration uuid?

> 
> Virt-handler doesn't have that mapping of a migration object to a pod, but the migration controller does.

hmm, I don't think I'm following.

What I'm trying to say that the destination virt-handler could just signal the target vmipod immediately when the vmi migration status becomes failed.
If there is a concern that we might signal a wrong pod, virt-handler could simply track the target pod when it calls `client.SyncMigrationTarget(vmi)`
and then when the status chances it would call `client.DestroyMigrationTarget(vmi)` (or add an env variable to the pod)

**Review by stu-gott — COMMENTED** [review]
/approve

**Comment by davidvossel** [comment]
> Do you think two target pods can start on the same node and be associate to the same vmi?

yeah, that shouldn't happen anymore. I address the situation where two target pods are active at the same time in this PR as well, so it shouldn't be a concern. 

> > What I'm trying to say is that the destination virt-handler could just signal the target vmipod immediately when the vmi migration status becomes failed.

I'll give it a shot.

I think i've been nervous about the reliability of having virt-handler do this, but as long as we're not allowing two migration target pods to be running in the cluster at the same time for the same VMI, this should at least be safe.

**Comment by davidvossel** [comment]
> I think i've been nervous about the reliability of having virt-handler do this, but as long as we're not allowing two migration target pods to be running in the cluster at the same time for the same VMI, this should at least be safe.

I spent some time trying to implement this at the handler level, and I'm still not confident. The timing of when the VMI hand off occurs to handler vs when/if the vmi MigrationState will get marked as failed vs all the different states some one can cancel a live migration in... (before handler handoff, after handler handoff but before migration is in flight, when migration is in flight... 

I can't keep the full picture in my head well enough to know if handler will can reliably send this signal. There are situations where the target handler won't even know the VMI exists yet (like someone cancelling the migration between migration scheduling and scheduled) But i know signalling the pod from the migration controller can be made reliable.

Maybe we should talk it through. There could be an angle i'm not seeing clearly.

**Comment by davidvossel** [comment]
/hold

Vladik and i talked about some tweaks here. I think we've figured out how to reliably signal early shutdown from handler. I'm going to give it another go.

**Comment by stu-gott** [comment]
/approve cancel

Will re-visit once signal reliability is re-worked.

**Comment by davidvossel** [comment]
Okay, I successfully removed the usage of the downward API and now have virt-handler signaling the target pod to exit early in the event of a failed migration.

I have a minor edge case left on this PR to fill.

If the migration fails due to the target pod exiting before the target pod ports have been published, the the migration controller should fail the migration on the vmi.Status.MigrationState object directly. That handles the scenario where the migration might get stuck if the target pod prep fails indefinitely.

**Comment by vladikr** [comment]
Everything looks good to me. I didn't find any obvious issue.
Just needs a rebase.
/approve

**Review by stu-gott — COMMENTED** [review]
/approve
/hold cancel

**Comment by stu-gott** [comment]
@davidvossel if I recall, you had added the hold label in order to get a chance to re-work the downward API. I've removed it as I don't think it's needed anymore. Please feel free to reverse this if you think the hold should still be in place.

**Comment by davidvossel** [comment]
/hold 

i want to take a look at one of the functional tests real quick today before this goes in.

**Comment by davidvossel** [comment]
/hold cancel

Okay, functional test addressed. This should be good to go.

**Comment by stu-gott** [comment]
@davidvossel looks like this PR got caught in the crossfire of another that modified live-migration-source.go

**Pull request merged**